lisa/tests/staging/utilclamp.py

```diff
@@ -0,0 +1,296 @@
+                bands,
+                key=lambda band: band[1] - band[0],
+                reverse=True
+            )
+            bands = bands[:nr_bands]
+            bands = sorted(bands, key=itemgetter(0))
+
+        return bands
+
+    @classmethod
+    def _get_phases(cls, plat_info):
+        capacities = cls._collect_capacities_flatten(plat_info)
+        bands = cls._get_bands(capacities)
+
+        def _band_mid(band):
+            return int((band[1] + band[0]) / 2)
+
+        """
+        Returns a list of phases. Each phases is described by a tuple:
+
+          (uclamp_val, util, (band))
+        """
+        return [(_band_mid(band),
+                 _band_mid(band) / 2,
+                 band)
+                for index, band in enumerate(bands)]
+
+    @classmethod
+    def get_rtapp_profile(cls, plat_info):
+        periods = [
+                Periodic(
+                    duty_cycle_pct=(util / PELT_SCALE) * 100,  # util to pct
+                    duration_s=5,
+                    period_ms=cls.TASK_PERIOD_MS,
+                    uclamp_min=uclamp_val,
+                    uclamp_max=uclamp_val)
+                for uclamp_val, util, expected in cls._get_phases(plat_info)
+        ]
+
+        return {'task': functools.reduce(lambda a, b: a + b, periods)}
+
+        return df
+
+    def _get_task_activation_capacity(self, task_name, start, end):
+
+        df_task = self.trace.analysis.tasks.df_task_activation(task_name)
+        df_task = df_task[['active', 'cpu']]
+
+        df_freq = self.trace.analysis.frequency.df_cpus_frequency()
+        df_freq = df_freq[['cpu', 'frequency']]
+        df_freq = df_freq.pivot(index=None,
+                                columns='cpu',
+                                values='frequency').reset_index()
+        df_freq.set_index('Time', inplace=True)
+
+        df = pd.concat([df_task, df_freq])
+        df = df.sort_index()
+
+        # Find the first activation
+        start = df[(df.index >= start) & (df.active == 1)].first_valid_index()
+
+        if (start >= end):
+            raise ValueError('Phase ends before it has even started')
+
+        df = df.ffill()
+        df = df[df.index >= start]
+        if (not np.isnan(end)):
+            df = df[df.index <= end]
+
+        # Fake Time, representing only the task activation period
+        df_add_delta(df, col='Time', inplace=True)
+        df = df.dropna()
+        df = df[df.active == 1]
+        df['Time'] = df['Time'].cumsum()
+
+        # Frequency to capacity
+        capacities = self._collect_capacities(self.plat_info)
+        for cpu in self.plat_info['nrg-model'].cpu_nodes:
+            cpu_id = cpu.cpus[0]
+            df[cpu_id] = df[cpu_id].map(capacities[cpu_id])
+
+        df['capacity'] = df.apply(lambda x: x[x.cpu], axis=1)
+        df['capacity_max'] = df.apply(
+                lambda x: max(capacities[x.cpu].values()), axis=1)
+
+        return [round(series_mean(df['capacity'], x=df['Time'])),
```

Comment:
Ahem ... yes, but I must admit that's completely empirical though.

With all the margins we have in the kernel, even with the very large band I used, we sometimes slightly exceed the maximum value. Rounding here prevent that problem, as the difference is always less than 1.

Reply:
If we only expect a difference of one, we should check for that explicitly at the spot the value is used. `round()` is the last thing we want to use for that, as it will not behave like "integer arithmetic rounding":
```
>>> round(0.4)
0
>>> round(0.6)
1
```